pep-0624.rst

```diff
@@ -0,0 +1,233 @@
+   ``_PyUnicode_TransformDecimalAndSpaceToASCII``
+   for converting from Unicode to numbers instead.
+
+
+Plan
+====
+
+Python 3.9
+----------
+
+Add ``Py_DEPRECATED(3.3)`` to these APIs.
+
+* ``PyUnicode_EncodeDecimal()``
+* ``PyUnicode_TransformDecimalToASCII()``.
+
+This change is committed already.
+https://github.com/python/cpython/commit/9c3840870814493fed62e140cfa43c2883e12181
+
+All other APIs have been marked ``Py_DEPRECATED(3.3)`` already.
+
+Python 3.10
+-----------
+
+Remove these APIs.
+
+* ``PyUnicode_Encode()``
+* ``PyUnicode_EncodeASCII()``
+* ``PyUnicode_EncodeLatin1()``
+* ``PyUnicode_EncodeUTF7()``
+* ``PyUnicode_EncodeUTF8()``
+* ``PyUnicode_EncodeUTF16()``
+* ``PyUnicode_EncodeUTF32()``
+* ``PyUnicode_EncodeUnicodeEscape()``
+* ``PyUnicode_EncodeRawUnicodeEscape()``
+* ``PyUnicode_EncodeCharmap()``
+* ``PyUnicode_TranslateCharmap()``
+
+
+Python 3.11
+-----------
+
+Remove these APIs.
+
+* ``PyUnicode_EncodeDecimal()``
+* ``PyUnicode_TransformDecimalToASCII()``
+
+
+Alternative ideas
+=================
+
+Make some private APIs public
+------------------------------
+
+``PyUnicode_EncodeUTF7()`` doesn't have public alternative APIs.
+
+Some APIs have alternative public APIs. But they are missing
+``const char *errors`` or ``int byteorder`` parameters.
+
+We can rename some private APIs and make them public to cover missing
+APIs and parameters.
+
+=============================== ================================
+ Rename to                       Rename from
+=============================== ================================
+``PyUnicode_EncodeASCII()``      ``_PyUnicode_AsASCIIString()``
+``PyUnicode_EncodeLatin1()``     ``_PyUnicode_AsLatin1String()``
+``PyUnicode_EncodeUTF7()``       ``_PyUnicode_EncodeUTF7()``
+``PyUnicode_EncodeUTF8()``       ``_PyUnicode_AsUTF8String()``
+``PyUnicode_EncodeUTF16()``      ``_PyUnicode_EncodeUTF16()``
+``PyUnicode_EncodeUTF32()``      ``_PyUnicode_EncodeUTF32()``
+=============================== ================================
+
+Pros:
+
+* We can have more consistent API set.
+
+Cons:
+
+* We have more public stable APIs to maintain.
+* Existing public APIs are enough for most use cases, and
+  ``PyUnicode_AsEncodedString()`` can be used in other cases.
+
+
+Replace ``Py_UNICODE *`` with ``Py_UCS4 *``
+-------------------------------------------
+
+We can replace ``Py_UNICODE`` (typedef of ``wchar_t``) with
+``Py_UCS4``. Since builtin codecs support UCS-4, we don't need to
+convert ``Py_UCS4 *`` string to Unicode object.
+
+
+Pros:
+
+* We can have more consistent API set.
+* User can encode UCS-4 string in C without creating Unicode object.
+
+Cons:
+
+* We have more public stable APIs to maintain.
+* Applications which uses UTF-8 or UTF-32 can not use these APIs
+  anyway.
+* Other Python implementations may not have builtin codec for UCS-4.
+* If we change the Unicode internal representation to UTF-8, we need
+  to keep UCS-4 support only for these APIs.
```

Comment:
Py_UNICODE is currently defined as wchar_t. So it's possible to use wide character literal L'c' and wide string literal L"string". It's not possible with Py_UCS4 which is a number, not a character.

Also, this change is likely to break existing C extensions relying on the assumption that Py_UNICODE is wchar_t.

Reply:
Of course, this is breaking change like removing deprecated API.